src/ipc-rpc-state-share.spec.ts

```diff
@@ -0,0 +1,161 @@
+      const hmacKey = Buffer.from('foobar');
+      const stream = new PassThrough();
+      stream.end(data);
+      try {
+        await readSignedStream(stream, 'somethingelse', hmacKey);
+        expect.fail('missed exception');
+      } catch (err: any) {
+        expect(err.message).to.include('Signature mismatch');
+      }
+    });
+  });
+
+  describe('RpcServer/RpcClient', function() {
+    let rpcServer: RpcServer;
+    let ExampleServer: typeof RpcServer & { create(): Promise<RpcServer> };
+    let handleRpcStub: sinon.SinonStub;
+
+    beforeEach(function() {
+      handleRpcStub = sinon.stub();
+      ExampleServer = class ExampleServer extends RpcServer {
+        handleRpc = handleRpcStub;
+
+        static async create() {
+          const instance = new this('example');
+          await instance._init();
+          return instance;
+        }
+      };
+    });
+
+    afterEach(async function() {
+      await rpcServer?.close();
+    });
+
+    it('allows communication over a local socket', async function() {
+      handleRpcStub.resolves({
+        cat: 'meow'
+      });
+
+      rpcServer = await ExampleServer.create();
+      const rpcClient = new RpcClient(rpcServer.handle);
+      const result = await rpcClient.makeRpcCall({ foo: 'bar' });
+      expect(result).to.deep.equal({ status: 200, cat: 'meow' });
+      expect(handleRpcStub).to.have.been.calledWith({ foo: 'bar' });
+    });
+
+    it('forwards errors from the handler', async function() {
+      handleRpcStub.rejects(new Error('something bad happened'));
+
+      rpcServer = await ExampleServer.create();
+      const rpcClient = new RpcClient(rpcServer.handle);
+      try {
+        await rpcClient.makeRpcCall({ foo: 'bar' });
+        expect.fail('missed exception');
+      } catch (err: any) {
+        expect(err.message).to.equal('something bad happened');
+      }
+    });
+  });
+
+  describe('StateShareServer/StateShareClient', async function() {
+    let fakeState: DevtoolsConnectionState;
+    let request: sinon.SinonStub;
+    let refresh: sinon.SinonStub;
+    let server: StateShareServer;
+    let client: StateShareClient;
+
+    beforeEach(function() {
+      request = sinon.stub();
+      refresh = sinon.stub();
+      fakeState = {
+        productName: 'MongoDB Something',
+        oidcPlugin: {
+          mongoClientOptions: {
+            authMechanismProperties: {
+              REFRESH_TOKEN_CALLBACK: refresh,
+              REQUEST_TOKEN_CALLBACK: request
+            }
+          }
+        }
+      } as unknown as DevtoolsConnectionState;
+      request.resolves({ accessToken: 'req-accesstoken' });
+      refresh.resolves({ accessToken: 'ref-accesstoken' });
+    });
+
+    afterEach(async function() {
+      await server.close();
+    });
+
+    it('can be used to share OIDC state', async function() {
+      server = await StateShareServer.create(fakeState);
+      client = new StateShareClient(server.handle);
```

Comment:
Is it something that we will do in vscode? or DataService will wrap this API? I am now thinking, in vscode, the connection controller uses DataService, but the language server uses CliServiceProvider to connect. We might need then to plan work to replace CliServiceProvider connection in LS with DataServie.

Reply:
> Is it something that we will do in vscode? 

Use this feature? Yes. Create these objects here manually? No, they’re implementation detials.

> or DataService will wrap this API? I am now thinking, in vscode, the connection controller uses DataService, but the language server uses CliServiceProvider to connect. We might need then to plan work to replace CliServiceProvider connection in LS with DataServie.

I don’t see an issue here. Both use devtools-connect to create the connection, so they should be compatible.

Reply:
We’re actually already relying on the fact that the options are compatible between them, I think: https://github.com/mongodb-js/vscode/blob/4d0c41d8d289b2ff66a5ec5c621f74d06a59177d/src/connectionController.ts#L833-L837